File: src/nightbot/sandbox.py
"""Docker sandbox for isolated code execution."""

from __future__ import annotations

import logging
import subprocess
import uuid
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional

logger = logging.getLogger(__name__)


@dataclass
class ExecResult:
    exit_code: int
    stdout: str
    stderr: str
    
    @property
    def success(self) -> bool:
        return self.exit_code == 0
    
    @property
    def output(self) -> str:
        """Combined output for LLM consumption."""
        parts = []
        if self.stdout.strip():
            parts.append(f"STDOUT:\n{self.stdout.strip()}")
        if self.stderr.strip():
            parts.append(f"STDERR:\n{self.stderr.strip()}")
        if not parts:
            return "(no output)"
        return "\n\n".join(parts)


class Sandbox:
    """Docker-based sandbox for running untrusted code."""
    
    def __init__(
        self,
        image: str = "nightbot-sandbox:latest",
        memory: str = "8g",
        cpus: int = 4,
        network: str = "host",
        workdir: str = "/workspace",
    ):
        self.image = image
        self.memory = memory
        self.cpus = cpus
        self.network = network
        self.workdir = workdir
        self.container_id: Optional[str] = None
        self.name = f"nightbot-sandbox-{uuid.uuid4().hex[:8]}"
    
    def create(self, mount_project: Optional[str] = None) -> str:
        """Create and start a sandbox container."""
        cmd = [
            "docker", "run", "-d",
            "--name", self.name,
            "--memory", self.memory,
            f"--cpus={self.cpus}",
            f"--network={self.network}",
            "-w", self.workdir,
        ]
        
        if mount_project:
            # Mount project read-only for context
            cmd.extend(["-v", f"{mount_project}:{self.workdir}:ro"])
            # Writable sandbox area
            cmd.extend(["-v", f"/tmp/{self.name}:/sandbox:rw"])
        
        cmd.extend([self.image, "sleep", "infinity"])
        
        logger.info(f"Creating sandbox: {self.name}")
        result = subprocess.run(cmd, capture_output=True, text=True)
        
        if result.returncode != 0:
            raise RuntimeError(f"Failed to create sandbox: {result.stderr}")
        
        self.container_id = result.stdout.strip()
        logger.info(f"Sandbox created: {self.container_id[:12]}")
        return self.container_id
    
    def exec(self, command: str, timeout: int = 120) -> ExecResult:
        """Execute a command inside the sandbox."""
        if not self.container_id:
            raise RuntimeError("Sandbox not created. Call create() first.")
        
        cmd = [
            "docker", "exec",
            "-w", "/sandbox",
            self.container_id,
            "bash", "-c", command,
        ]
        
        logger.debug(f"Sandbox exec: {command[:100]}...")
        
        try:
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=timeout,
            )
            return ExecResult(
                exit_code=result.returncode,
                stdout=result.stdout[-5000:],  # truncate to last 5k chars
                stderr=result.stderr[-5000:],
            )
        except subprocess.TimeoutExpired:
            return ExecResult(
                exit_code=-1,
                stdout="",
                stderr=f"Command timed out after {timeout}s",
            )
    
    def write_file(self, path: str, content: str) -> ExecResult:
        """Write a file inside the sandbox."""
        # Escape for bash
        escaped = content.replace("'", "'\\''")
        return self.exec(f"cat > /sandbox/{path} << 'NIGHTBOT_EOF'\n{content}\nNIGHTBOT_EOF")
    
    def destroy(self) -> None:
        """Stop and remove the sandbox container."""
        if not self.container_id:
            return
        
        logger.info(f"Destroying sandbox: {self.name}")
        subprocess.run(
            ["docker", "rm", "-f", self.container_id],
            capture_output=True,
        )
        self.container_id = None
    
    def __enter__(self):
        self.create()
        return self
    
    def __exit__(self, *args):
        self.destroy()


def build_sandbox_image():
    """Build the default sandbox Docker image."""
    dockerfile = """\
FROM node:20-slim

RUN apt-get update && apt-get install -y \\
    git curl python3 python3-pip build-essential \\
    && rm -rf /var/lib/apt/lists/*

RUN npm install -g typescript ts-node vitest jest pnpm yarn

WORKDIR /sandbox
"""
    
    tmp = Path("/tmp/nightbot-sandbox-build")
    tmp.mkdir(exist_ok=True)
    (tmp / "Dockerfile").write_text(dockerfile)
    
    logger.info("Building sandbox image...")
    result = subprocess.run(
        ["docker", "build", "-t", "nightbot-sandbox:latest", str(tmp)],
        capture_output=True,
        text=True,
    )
    
    if result.returncode != 0:
        raise RuntimeError(f"Failed to build sandbox image: {result.stderr}")
    
    logger.info("Sandbox image built successfully")

File: src/sandbox.ts
import { execSync } from "node:child_process";
import { randomUUID } from "node:crypto";

export interface ExecResult {
  exitCode: number;
  stdout: string;
  stderr: string;
}

export class Sandbox {
  private containerId: string | null = null;
  public readonly name: string;

  constructor(
    private image = "nightbot-sandbox:latest",
    private memory = "8g",
    private cpus = 4,
    private network = "host",
  ) {
    this.name = `nightbot-${randomUUID().slice(0, 8)}`;
  }

  create(mountProject?: string): string {
    const args = [
      "docker", "run", "-d",
      "--name", this.name,
      "--memory", this.memory,
      `--cpus=${this.cpus}`,
      `--network=${this.network}`,
      "-w", "/sandbox",
    ];

    if (mountProject) {
      args.push("-v", `${mountProject}:/workspace:ro`);
    }

    args.push(this.image, "sleep", "infinity");

    const id = execSync(args.join(" "), { encoding: "utf-8" }).trim();
    this.containerId = id;
    return id;
  }

  exec(command: string, timeout = 120): ExecResult {
    if (!this.containerId) throw new Error("Sandbox not created");

    try {
      // Pipe command to bash stdin so heredocs and multi-line scripts work correctly
      const stdout = execSync(
        `docker exec -i ${this.containerId} bash`,
        {
          input: command,
          encoding: "utf-8",
          timeout: timeout * 1000,
          maxBuffer: 10 * 1024 * 1024,
        },
      );
      return { exitCode: 0, stdout: stdout.slice(-5000), stderr: "" };
    } catch (err: unknown) {
      const e = err as { status?: number; stdout?: string; stderr?: string };
      return {
        exitCode: e.status ?? 1,
        stdout: (e.stdout ?? "").slice(-5000),
        stderr: (e.stderr ?? "").slice(-5000),
      };
    }
  }

  writeFile(path: string, content: string): ExecResult {
    const b64 = Buffer.from(content).toString("base64");
    return this.exec(`mkdir -p $(dirname /sandbox/${path}) && echo ${b64} | base64 -d > /sandbox/${path}`);
  }

  writeFileAt(absolutePath: string, content: string): ExecResult {
    const b64 = Buffer.from(content).toString("base64");
    return this.exec(`mkdir -p $(dirname ${absolutePath}) && echo ${b64} | base64 -d > ${absolutePath}`);
  }

  readFile(absolutePath: string): string {
    const result = this.exec(`cat ${absolutePath} 2>/dev/null`);
    return result.stdout;
  }

  listFiles(dir: string, pattern = "*"): string[] {
    const result = this.exec(`find ${dir} -name "${pattern}" -type f 2>/dev/null | head -50`);
    return result.stdout.trim().split("\n").filter(Boolean);
  }

  createWithRepo(repoPath: string): string {
    const id = this.create(repoPath);
    this.exec("cp -r /workspace /project");
    this.exec("cd /project && (grep -q node_modules .gitignore 2>/dev/null || echo -e 'node_modules/\\ndist/\\ncoverage/' >> .gitignore)");
    this.exec("cd /project && git add -A 2>/dev/null; git stash 2>/dev/null || true");
    return id;
  }

  extractDiff(): string {
    if (!this.containerId) throw new Error("Sandbox not created");
    this.exec("cd /project && git add -A 2>/dev/null");
    const result = this.exec("cd /project && git diff --cached HEAD -- . ':!node_modules' ':!package-lock.json' ':!pnpm-lock.yaml'");
    return result.stdout;
  }

  extractChangedFiles(): Array<{ path: string; content: string }> {
    if (!this.containerId) throw new Error("Sandbox not created");
    this.exec("cd /project && git add -A 2>/dev/null");
    const result = this.exec("cd /project && git diff --cached --name-only HEAD -- . ':!node_modules' ':!package-lock.json' ':!pnpm-lock.yaml'");
    const files = result.stdout.trim().split("\n").filter(Boolean);
    return files.filter(f => !f.startsWith("node_modules/")).map(f => {
      const content = this.exec(`cat /project/${f}`);
      return { path: f, content: content.stdout };
    });
  }

  destroy(): void {
    if (!this.containerId) return;
    try {
      execSync(`docker rm -f ${this.containerId}`, { encoding: "utf-8" });
    } catch {}
    this.containerId = null;
  }
}

export function buildSandboxImage(): void {
  const dockerfile = `FROM node:20-slim
RUN apt-get update && apt-get install -y git curl python3 build-essential && rm -rf /var/lib/apt/lists/*
RUN npm install -g typescript tsx pnpm
WORKDIR /sandbox`;

  execSync(`echo '${dockerfile}' | docker build -t nightbot-sandbox:latest -`, {
    encoding: "utf-8",
    stdio: "inherit",
  });
}
